File: examples/memory/typescript/test-contracts.ts
import { ArtifactView, buildContext, GovernedMemory, MemoryRecord } from "./contracts.js"

const assert = (condition: boolean, message: string): void => { if (!condition) throw new Error(message) }
const throws = (fn: () => void): void => {
  let failed = false
  try { fn() } catch { failed = true }
  assert(failed, "expected an exception")
}

const memory = new GovernedMemory()
const preference: MemoryRecord = { key: "preference/theme", content: "dark", tenantId: "t1", subjectId: "u1", provenance: "user_input", trust: "user_confirmed", consent: false }
throws(() => memory.put(preference))
memory.put({ ...preference, consent: true })
assert(memory.get("preference/theme", "t2", "u1") === undefined, "tenant isolation failed")
assert(memory.get("preference/theme", "t1", "u1")?.content === "dark", "memory lookup failed")
memory.put({ key: "temporary", content: "old", tenantId: "t1", subjectId: "u1", provenance: "tool_result", trust: "verified", consent: false, retentionUntil: Date.now() / 1000 - 1 })
assert(memory.get("temporary", "t1", "u1") === undefined, "retention failed")
const context = buildContext("rules", ["old history", "latest history"], [{ artifactId: "a1", summary: "report", mediaType: "text/plain", checksum: "abc" } as ArtifactView], 30)
assert(JSON.stringify(context.slice(0, 2)) === JSON.stringify(["rules", "[artifact:a1 text/plain sha256=abc] report"]), "artifact view failed")
assert(context.at(-1) === "latest history", "history selection failed")
console.log("memory/context contract tests: OK")

File: examples/memory/python/test_contracts.py
import time
import unittest

from contracts import ArtifactView, GovernedMemory, MemoryRecord, build_context


class MemoryContextContractTest(unittest.TestCase):
    def test_memory_is_scoped_and_consent_gated(self):
        memory = GovernedMemory()
        with self.assertRaises(PermissionError):
            memory.put(MemoryRecord("preference/theme", "dark", "t1", "u1", "user_input", "user_confirmed", False))
        memory.put(MemoryRecord("preference/theme", "dark", "t1", "u1", "user_input", "user_confirmed", True))
        self.assertIsNone(memory.get("preference/theme", tenant_id="t2", subject_id="u1"))
        self.assertEqual(memory.get("preference/theme", tenant_id="t1", subject_id="u1").content, "dark")

    def test_retention_forget_and_context_budget_are_deterministic(self):
        memory = GovernedMemory()
        memory.put(MemoryRecord("temporary", "old", "t1", "u1", "tool_result", "verified", False, time.time() - 1))
        self.assertIsNone(memory.get("temporary", tenant_id="t1", subject_id="u1"))
        self.assertTrue(memory.forget("missing", tenant_id="t1", subject_id="u1") is False)
        context = build_context("rules", ["old history", "latest history"], [ArtifactView("a1", "report", "text/plain", "abc")], token_budget=30)
        self.assertEqual(context[:2], ["rules", "[artifact:a1 text/plain sha256=abc] report"])
        self.assertEqual(context[-1], "latest history")


if __name__ == "__main__":
    unittest.main()
